core/index/directory_reader_impl.cpp

```diff
@@ -0,0 +1,275 @@
+
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2022 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Andrey Abramov
+////////////////////////////////////////////////////////////////////////////////
+
+#include "directory_reader_impl.hpp"
+
+#include "utils/directory_utils.hpp"
+
+#include <absl/container/flat_hash_map.h>
+#include <absl/strings/str_cat.h>
+
+namespace irs {
+
+struct DirectoryReaderImpl::Init {
+  Init(const directory& dir, const DirectoryMeta& meta);
+
+  FileRefs file_refs;
+  uint64_t docs_count{};
+  uint64_t live_docs_count{};
+};
+
+DirectoryReaderImpl::Init::Init(const directory& dir,
+                                const DirectoryMeta& meta) {
+  const bool has_segments_file = !meta.filename.empty();
+
+  auto& [file_refs, docs_count, live_docs_count] = *this;
+
+  file_refs.reserve(meta.index_meta.segments.size() +
+                    size_t{has_segments_file});
+
+  auto& refs = dir.attributes().refs();
+  for (const auto& [filename, segment] : meta.index_meta.segments) {
+    file_refs.emplace_back(refs.add(filename));
+    docs_count += segment.docs_count;
+    live_docs_count += segment.live_docs_count;
+  }
+
+  if (has_segments_file) {
+    file_refs.emplace_back(refs.add(meta.filename));
+  }
+}
+
+namespace {
+
+MSVC_ONLY(__pragma(warning(push)))
+MSVC_ONLY(__pragma(warning(disable : 4457)))  // variable hides function param
+index_file_refs::ref_t LoadNewestIndexMeta(IndexMeta& meta,
+                                           const directory& dir,
+                                           format::ptr& codec) noexcept {
+  // if a specific codec was specified
+  if (codec) {
```

Comment:
Code for gettting "ref" seems to be the same for both branches. Maybe extract it as lambda?

Reply:
https://github.com/iresearch-toolkit/iresearch/pull/447#discussion_r1068372005

Reply:
Let's address it in a separate PR.